cli/cmd/edges.go

```diff
@@ -0,0 +1,360 @@
+package cmd
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"sort"
+	"strings"
+	"text/tabwriter"
+
+	"github.com/linkerd/linkerd2/controller/api/util"
+	pb "github.com/linkerd/linkerd2/controller/gen/public"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+type edgesOptions struct {
+	namespace    string
+	outputFormat string
+}
+
+func newEdgesOptions() *edgesOptions {
+	return &edgesOptions{
+		namespace:    "",
+		outputFormat: tableOutput,
+	}
+}
+
+type indexedEdgeResults struct {
+	ix   int
+	rows []*pb.Edge
+	err  error
+}
+
+func newCmdEdges() *cobra.Command {
+	options := newEdgesOptions()
+
+	cmd := &cobra.Command{
+		Use:   "edges [flags] (RESOURCETYPE)",
+		Short: "Display connections between resources, and Linkerd proxy identities",
+		Long: `Display connections between resources, and Linkerd proxy identities.
+
+  The RESOURCETYPE argument specifies the type of resource to display edges within. A namespace must be specified.
+
+  Examples:
+  * deploy
+  * ds
+  * job
+  * po
+  * rc
+  * sts
+
+  Valid resource types include:
+  * daemonsets
+  * deployments
+  * jobs
+  * pods
+  * replicationcontrollers
+  * statefulsets`,
+		Example: `  # Get all edges between pods in the test namespace.
+  linkerd edges po -n test`,
+		Args:      cobra.ExactArgs(1),
+		ValidArgs: util.ValidTargets,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			reqs, err := buildEdgesRequests(args, options)
+			if err != nil {
+				return fmt.Errorf("Error creating edges request: %s", err)
+			}
+
+			// The gRPC client is concurrency-safe, so we can reuse it in all the following goroutines
+			// https://github.com/grpc/grpc-go/issues/682
+			client := checkPublicAPIClientOrExit()
+			c := make(chan indexedEdgeResults, len(reqs))
+			for num, req := range reqs {
+				go func(num int, req *pb.EdgesRequest) {
```

Comment:
these concurrent requests are intended for multiple-resource commands like `linkerd edges deploy/ pod/`. if we're planning to support that at some point, i'm ok leaving it here. if not, better to just have `buildEdgesRequests` return a single request, and make that request synchronously.

Reply:
Yes, planning on supporting multiple-resource commands in the next iteration of this feature, so going to leave this in for now.